client/driver/fabfile.py

```diff
@@ -1029,3 +1029,98 @@ def task_status_ui_test():
+        summary = json.load(fin)
+        summary['start_time'] = i * 20000000
+        summary['observation_time'] = 100
+        summary['end_time'] = (i + 1) * 20000000
+    with open('./integrationTests/data/x__summary.json', 'w') as fout:
+        json.dump(summary, fout)
+    return gain
+
+
+@task
+def integration_tests():
+    # Create test website
+    response = requests.get(dconf.WEBSITE_URL + '/test/create/')
+    LOG.info(get_content(response))
+
+    # Upload training data
+    LOG.info('Upload training data to no tuning session')
+    upload_batch(result_dir='./integrationTests/data/', upload_code='ottertuneTestNoTuning')
+
+    # periodic tasks haven't ran, lhs result returns.
+    LOG.info('Test no pipeline data, LHS returned')
+    upload_result(result_dir='./integrationTests/data/', prefix='1__',
+                  upload_code='ottertuneTestTuningGPR')
+    response = get_result(upload_code='ottertuneTestTuningGPR')
+    assert response['status'] == 'lhs'
+
+    # wait celery periodic task finishes
+    assert wait_pipeline_data_ready(), "Pipeline data failed"
+
+    total_n = 20
+    last_n = 10
+
+    max_gain = 0
+    simulate_db_run(1, {'effective_cache_size': '0kB'})
+    for i in range(2, total_n + 2):
+        LOG.info('Test GPR (gaussian process regression)')
+        upload_result(result_dir='./integrationTests/data/', prefix='x__',
+                      upload_code='ottertuneTestTuningGPR')
+        response = get_result(upload_code='ottertuneTestTuningGPR')
+        assert response['status'] == 'good'
+        gain = simulate_db_run(i, response['recommendation'])
+        if max_gain < gain:
+            max_gain = gain
+        elif i > total_n - last_n + 2:
+            assert gain > max_gain / 2.0
+
+    max_gain = 0
+    simulate_db_run(1, {'effective_cache_size': '0kB'})
+    for i in range(2, total_n + 2):
+        LOG.info('Test DNN (deep neural network)')
+        upload_result(result_dir='./integrationTests/data/', prefix='x__',
+                      upload_code='ottertuneTestTuningDNN')
+        response = get_result(upload_code='ottertuneTestTuningDNN')
+        assert response['status'] == 'good'
+        gain = simulate_db_run(i, response['recommendation'])
+        if max_gain < gain:
+            max_gain = gain
+        elif i > total_n - last_n + 2:
+            assert gain > max_gain / 2.0
+
+    max_gain = 0
+    simulate_db_run(1, {'effective_cache_size': '0kB'})
+    for i in range(2, total_n + 2):
+        upload_result(result_dir='./integrationTests/data/', prefix='x__',
+                      upload_code='ottertuneTestTuningDDPG')
+        response = get_result(upload_code='ottertuneTestTuningDDPG')
+        assert response['status'] == 'good'
+        gain = simulate_db_run(i, response['recommendation'])
+        if max_gain < gain:
+            max_gain = gain
+        elif i > total_n - last_n + 2:
+            assert gain > max_gain / 2.0
```

Comment:
Is the condition too strict ?  Sometimes one bad recommendation is fine for exploration. 
Will the test always passes? They may have different random seeds for different runs.

Reply:
I have been thinking through the condition and testing some options locally and I finally think this condition is the most balanced one. As far as I know, the test will always pass given almost any random seed.
The reason is that the function of the gain and the knobs is quite straightforward, as the gain is only related to one knob. And all of the three algorithms converge on that knob basically after 5 iterations. Even if it explores, the knob will not be smaller than half of the max value it has ever seen.
If in the future you want to encourage exploration and this condition is broken, you can come back to change the test.